lib/front_desk.rb

```diff
@@ -0,0 +1,105 @@
+require_relative 'reservation'
+require_relative 'date_range'
+
+module Hotel
+  class FrontDesk
+    attr_accessor :reservations, :rooms, :calendar
+
+    def initialize
+      @rooms = (1..20).map { |i| i }
+      @reservations = []
+      @calendar = {}
+    end
+
+    def populate_calendar(reservation)
+      temp = reservation.date_range.dates
+      temp.pop
+      temp.each do |date|
+        if !@calendar.key?(date)
+          @calendar[date] =[]
+        end
+        @calendar[date] << reservation
+      end
+      return @calendar
+    end
+
+    # User: I can access the list of all of the rooms in the hotel
+    def list_rooms
+      return @rooms
+    end
+
+    # User: I can make a reservation of a room for a given date range, and that room will not be part of any other reservation overlapping that date range
+    def reserve_room(requested_dates)
+       available_rooms = available_rooms(requested_dates)
+       
+       book_room = available_rooms.first
+       new_reservation = Hotel::Reservation.new(
+        room: book_room, 
+        date_range: requested_dates,
+      )
+      @reservations.push(new_reservation)
+      populate_calendar(new_reservation)
+      return new_reservation
+    end
+
+    # User: I can access the list of reservations for a specific date, so that I can track reservations by date
+    def date_reservations(date)
+      date_resv = @calendar[date]
+      return date_resv
+    end
+
+    # User: I access the list of reservations for a specified room and a given date range
+    # I split this user story into a handful of separate methods to perform different parts of the request
+
+    # returns an array of reservations for requested dates
+    def range_reservations(requested_dates)
+      range_resv = []
+      requested_dates.dates.each do |date|
+        if @calendar.key?(date)
+          range_resv.push(*@calendar[date])
+        end
+      end
+      range_resv = range_resv.uniq
+  
+      return range_resv
+    end
```

Comment:
You can clean this up using `select` and `flat_map`:

```suggestion
    def range_reservations(requested_dates)
      return requested_dates.dates.select do |date|
        @calendar.key?(date)
      end.flat_map do |date|
          @calendar[date]
      end.uniq
    end
```

(Since it's kind of obscure: `flat_map` is equivalent to doing a `map` followed by a `flatten`.  It basically says "go through the input and produce an array out of the little arrays this block outputs".)